src/graph_ir.rs

```diff
@@ -305,12 +307,56 @@ impl GraphIR {
                     input_buffers: sum_inputs,
                     output_buffer: sum_output,
                 });
-                // This node's input buffer is the sum output buffer
-                assignments.push(sum_output);
-                allocator.release(sum_buffer);
+                // This node's input buffer is the sum output buffer. Release it once the node
+                // assignments are done.
+                input_buffers.push(sum_output);
+                buffers_to_release.push(sum_buffer);
             }
         }
 
+        for port in node.outputs.iter() {
+            let type_index = port.type_idx;
+            let edges = adjacent_edges
+                .outgoing
+                .iter()
+                .filter(|edge| edge.src_port == port.id)
+                .collect::<Vec<_>>();
+
+            if edges.is_empty() {
+                // Case 5: The port is an output and it is unconnected. Acquire a buffer and
+                //         assign it. The buffer does not need to be cleared. Release the
+                //         buffer once the node assignments are done.
+                let buffer = allocator.acquire(type_index);
+                output_buffers.push(BufferAssignment {
+                    buffer_index: buffer.idx,
+                    generation: buffer.generation,
+                    type_index: buffer.type_idx,
+                    port_id: port.id,
+                    should_clear: false,
+                });
+                buffers_to_release.push(buffer);
+            } else {
+                // Case 6: The port is an output. Acquire a buffer, and add to the assignment
+                //         table with any corresponding edge IDs. For each edge, update the
+                //         assigned buffer table. Buffer should not be cleared or released.
+                let buffer = allocator.acquire(type_index);
+                for edge in &edges {
+                    assignment_table.insert(edge.id, buffer.clone());
+                }
+                output_buffers.push(BufferAssignment {
+                    buffer_index: buffer.idx,
+                    type_index: buffer.type_idx,
+                    generation: buffer.generation,
+                    port_id: port.id,
+                    should_clear: false,
+                });
+            }
+        }
+
+        for buffer in buffers_to_release.drain(..) {
```

Comment:
You remove the `drain` call here, I think

Reply:
This uses `drain` to avoid needing to clone the `Rc`s before invoking the `release()` method.

Reply:
👍